### DISCUSSION TITLE
Fanbased Voice acting

### DISCUSSION THREAD

**Discussion by AphexSign** [exult/exult#410] [General]
Can you implement voice-acting support to Exult for a future possibility fanbased nonprofit voice acting? 

Let's say for example:
Function from usecode 0827, string number 100 ""Through his speeches and writings he hath changed the lives of many people, including mine own."" plays sound "si/voice/0827_100.ogg"(if sound file is present)

Thanks in advance!

**Comment by grazelandsnomad** [comment]
I'm looking forward to this as well. But generating it with Elevenlabs AI. And translating the game to spanish.

**Reply by grazelandsnomad** [reply]
I will give an example, wait.

**Reply by grazelandsnomad** [reply]
![image](https://github.com/exult/exult/assets/45631468/f0811957-1670-4959-b485-93a24e3f8aa8)
https://drive.google.com/file/d/1tI3boE8JCHP6FGPDci45IZ_WppGCje0r/view?usp=sharing

**Comment by grazelandsnomad** [comment]
![image](https://github.com/exult/exult/assets/45631468/1287856e-162d-4a51-8e69-a49f3060f9c2)

https://drive.google.com/file/d/1tLTZv8wmpGiWyFWATmus0LZuoOzNwqww/view?usp=drive_link
https://drive.google.com/file/d/1tP3ImnIU-M-PgyhxaEVy0078OJoyPb0L/view?usp=drive_link

**Comment by wench** [comment]
Dialog in Exult does not work like that. The Usecode doesn't use string indices to display dialog, it passes actual strings to the engine. Using string loookups instead of indices for speech is certainly possible (it's how Speech works in Ultima8 and we implemented it in Pentagram) but it has certain problems because the dialog strings are not constant, they are dynamic, the usecode can construct dialog strings dynamically by doing things like including the Avatar's name from character creation in dialog, and gendered word substitutions (eg M'lord and M'Lady). Ultima 8 solved the Avatar name problem by just not using it in dialog with speech. If you want to add speech to all dialog then you need a solution to the dynamic dialog problem.

IMO the best way to add speech will be some engine changes so usecode can specify speech files for normal dialog and modify the usecode so it tells the engine what voice file to use for every line of dialog. For engine changes an actual workable proposal would be needed first before we would make any changes.

**Comment by grazelandsnomad** [comment]
Thanks for your response.
Maybe, if the player decides using a voice mod like the one proposed… I could overwrite all strings that use Avatar's name (rolling over it) and all gendered expressions for neutral ones. Except of singled M'Lord or M'Lady (if used as one-liner), to simplify the matter. I know it's not the ideal solution, but… I don't know. Maybe it would be worth losing a bit of fidelity in exchange for the voiced game. At least for me, it would, and for some players certainly would. And would help gain new younger players, possibly. Which they are not that used to actually read. (Sad but true, you know). But overwriting all the strings code just for this purpose I understand is overkill. There's no way to get the id for a certain string? Even if it will be constructed on the fly? And rolling over it with the voice? Perhaps the text could display the gendered and M'Lady etc. and the voice just skip the gendered parts?

**Comment by AphexSign** [comment]
If I remember correctly, in Ultima 6 Fm towns (full voice acting included), this problem was solved as follows:
1) The spell shop or provisioners provided a simplified version of the dialogue.
2) The hero's names were not pronounced. Most sentences were gender-neutral.

**Comment by AphexSign** [comment]
Bump a thread. I made a small sample of the voice acting of Lord British and narrator.

https://dropmefiles.com/aXhbB

password for archive: 123 
I respect the voices of RG (Lord British) and the Narrator-actor. Therefore, if it becomes possible to implement the voice acting of Ultima 7. I am considering replacing their voices with others so as not to violate their rights.

Is it possible to add a function in Exult for Usecode to play a custom sound effect? 
For voice acting. I have a plan to manually modify each character script file so that before each line spoken by the character, the function play_sound_effect("010C_0001.ogg") is called. 
If the voiceover file is missing, the game will not crash, as there will be an exception handling that either plays "template_sound.ogg" with silence.

If we recall the game Faery Tale Adventure II: Halls of the Dead, during the dialogue with the merchant, this is exactly what happens. The merchant sells an item:
"I am selling you + name-of-the-item + for + amount-of-coins + coins."
That is, one line could contain 5 links to different inner sound files.

**Comment by marzojr** [comment]
*Adding* a way to play such audio would be easy. Making practical use of it would not be.

As @wench says, U7 dialog doesn't work like that. There are a lot of dynamically computed strings that are combined together in usecode before being pushed to the engine for display. At the point where the string is actually pushed to the engine for display (which is the point where you would want to play the audio), there is usually no information on how it was assembled -- just the combined result.

So instead of having something like

```dart
say("I am selling you ",
    "a",
    " ",
    "sword",
    "",
    " for ",
    price,
    " ",
    "coins",
    ".");
```

You have

```dart
var coinSuffix;
if (price != 1) {
    coinSuffix = "s";
} else {
    coinSuffix = "";
}
var itemArticle, itemPluralSuffix;
if (quantity != 1) {
    itemArticle = quantity + itemArticlePlural[index];
    itemPluralSuffix = itemPluralSuffix[index];
} else {
    itemArticle = itemArticleSingular[index];
    itemPluralSuffix = "";
}
say("I am selling you ",
    itemArticle,
    " ",
    itemName[index],
    itemPluralSuffix,
    " for ",
    itemPrice[index],
    " ",
    "coin",
    coinSuffix ,
    ".");
```

This is slightly simplified, but it shows the point: you don't know even at the end if you are saying "coin" or "coins", or whether to say plural or singular for the item name, or what article to use. Also note that `itemArticle` is a combination of two strings which you can't easily untangle.

It can also get even more complicated: for example, for garlic, `itemArticlePlural[index]` can be `" cloves of"`, while `itemArticleSingular[index]` can be `a clove of`.

To make actual voice acting for U7, you would basically have to edit all dialog to keep track of what pieces of audio (and which order!) will be needed for the speech. I don't think this is very realistic.

I also find it unrealistic to expect that the tone and cadence among all those disparate pieces of audio would be consistent in all cases.

**Comment by wench** [comment]
Being pedantic here, a play_sound_effect() function would be inappropriate for playing speech. Exult specifically separates sound effects and speech so you can do things like have separate independent global volume levels for Speech and Sound Effects.

**Comment by AphexSign** [comment]
As a compromise, not all of the NPC's lines can be voiced.
If phrase is heavily scripted, then you can skip it. Like it was in Fm-towns version of Ultima 6. 
They simply did not voice some of the phrases or voiced them in general phrases.

I see trainers, magicians, and merchants using such complicated phrases.
But even they have lines without complicated script construction, such as a "name" or "job".

I wanted Exult to have a function that has access to a specific folder/archive where *.ogg files are stored
and can play audio files. In addition, a function would be useful to which, as an argument, it could be passed 
a list of links to files in case of complex voice-script logic.

Something like that:

    std::vector<std::string> soundEffects = {
        "lb/001.ogg",
        "lb/002.ogg",
        "lb/003.ogg"
	...
	etc
    };
    
    playSoundEffects(soundEffects);

**Reply by KnightCaptainU7** [reply]
I think you'd have to edit the Usecode to add these in, short of an Exult engine change.
